src/lc3/vm.rs

```diff
@@ -0,0 +1,806 @@
+        let store_register: usize = register_index.into();
+        *self
+            .registers
+            .get_mut(store_register)
+            .ok_or(VMError::InvalidRegister)? = value;
+        Ok(())
+    }
+
+    /// Update flags based on the content of the register requested
+    ///
+    /// If register is not 0 to 7, an Error is returned
+    fn update_flags(&mut self, register: u16) -> Result<(), VMError> {
+        let value = *self.get_register(register)?;
+        if value == 0x0 {
+            self.cond = ConditionFlag::Zro;
+        } else if value >> 15 == 1 {
+            self.cond = ConditionFlag::Neg;
+        } else {
+            self.cond = ConditionFlag::Pos;
+        }
+        Ok(())
+    }
+
+    pub fn read_image(&mut self, image_path: &str) -> Result<(), VMError> {
+        self.memory.read_image(image_path)
+    }
+
+    fn mem_read(&mut self, index: usize) -> Result<u16, VMError> {
+        self.memory.mem_read(index)
+    }
+
+    pub fn mem_write(&mut self, value: u16, index: usize) -> Result<(), VMError> {
+        self.memory.mem_write(value, index)
+    }
+
+    /// Main execution loop
+    ///
+    /// Read instruction, increments the PC, execute the instruction
+    /// Repeat this while the machine is in a state of running
+    pub fn run(&mut self) -> Result<(), VMError> {
+        // start machine
+        self.running = true;
+
+        while self.running {
+            // Fetch
+            let instr = self.mem_read(self.pc.into())?;
+            // Increment PC
+            self.pc = self
+                .pc
+                .checked_add(1)
+                .ok_or(VMError::MemoryIndex(String::from("PC out of bounds")))?;
+            // Execute
+            self.execute(instr)?;
+        }
+
+        Ok(())
+    }
+
+    /// Based on the opcode from the instruction bits, executes the corresponding procedure
+    fn execute(&mut self, instr: u16) -> Result<(), VMError> {
+        let op: Opcode = (instr >> 12).try_into()?;
+
+        match op {
+            Opcode::BR => self.br(instr),
+            Opcode::Add => self.add(instr),
+            Opcode::LD => self.ld(instr),
+            Opcode::ST => self.st(instr),
+            Opcode::Jsr => self.jsr(instr),
+            Opcode::And => self.and(instr),
+            Opcode::Ldr => self.ldr(instr),
+            Opcode::Str => self.str(instr),
+            Opcode::Rti => Err(VMError::InvalidOpcode),
+            Opcode::Not => self.not(instr),
+            Opcode::Ldi => self.ldi(instr),
+            Opcode::Sti => self.sti(instr),
+            Opcode::Jmp => self.jmp(instr),
+            Opcode::Res => Err(VMError::InvalidOpcode),
+            Opcode::Lea => self.lea(instr),
+            Opcode::Trap => self.trap(instr),
+        }
+    }
+
+    /// ADD
+    ///
+    /// ADD DR, SR1, SR2 or ADD DR, SR1, imm5
+    ///
+    /// Stores in DR the addition between SR1 and SR2 or SR1 and imm5 sign extended
+    /// Bit 5 if it is 1, then is read as a imm5
```

Comment:
In this documentation, `Bit 5`, `SR2` and `imm5` are not related. Please refactor this so that the relationship is explicit. For example:

```suggestion
    /// If Bit 5 is 0, stores in DR the addition between SR1 and SR2.
    /// If Bit 5 is 1, stores in DR the addition between SR1 and imm5.
```

Reply:
Done, [b5d8659](https://github.com/tomip01/lc3-vm/pull/6/commits/b5d8659e63627c95f856df37621ae7fab1d53161)